content/ui-testing-handbook/react/en/visual-testing.md

```diff
@@ -0,0 +1,313 @@
+
+### 2. Does it <i>still</i> look right?
+
+You built it right the first time. It looks good in _all states_. But changes happen over the natural course of development. Bugs inevitably sneak in. This is especially true for interfaces. A minor CSS tweak can break a component or one of its states.
+
+You can’t manually check the breadth of the UI every time you make a change. You need something more automated.
+
+## Visual testing
+
+Visual testing allows you to tackle both these tasks with one unified workflow. It is the process of verifying the appearance of a component as you’re building it. And again as you iterate to ship new features.
+
+Here's what the visual testing workflow looks like:
+
+1.  🏷 **Isolate** components. Use [Storybook](https://storybook.js.org/) to focus on and test one component at a time.
+2.  ✍🏽 **Write out the test cases.** Each state is reproduced using props and mock data.
+3.  🔍 **Manually verify** the appearance of each test case.
+4.  📸 **Catch UI bugs automatically.** Capture snapshots of each test case and use machine-based diffing to check for regressions.
+
+The crux of visual testing is isolating the UI from the rest of the app (data, backend, APIs). That allows you to observe each state individually. You can then manually spot check and automatically regression test those states.
+
+Let's go through each step in detail.
+
+### 1. Isolate components
+
+It’s much easier to pinpoint bugs by testing one component at a time and writing a test case for each of its states. The conventional approach is to build the component on the application page where it is first used. Which makes it hard to simulate and verify all these states. There’s a better way—Storybook.
+
+Storybook is the industry-standard for building components in isolation. It’s used by Twitter, Slack, Airbnb, Shopify, Stripe, and Microsoft. It is packaged as a small standalone tool that lives alongside your app, giving you:
+
+- 📥 A **sandbox** to render each component in isolation
+- 🔭 Visualize all its **states** as _stories_
+- 📑 **Document** props and usage guidelines for each component
+- 🗃️ A **directory** of all your components to make discovery easier
+
+Let’s go back to that Task component. To isolate it means that we load up and render this one component by itself. For that, we need Storybook.
+
+### Setup Storybook
+
+Our project is preconfigured to use Storybook. The config lives in the `.storybook` folder and all the necessary scripts have been added to `package.json`.
+
+We can start by creating a story file for the Task component. This registers the component with Storybook and adds in one default test case.
+
+```javascript:title=src/components/Task.stories.js
+import React from 'react';
+import { Task } from './Task';
+
+export default {
+  component: Task,
+  title: 'Task',
+  argTypes: {
+    onArchiveTask: { action: 'onArchiveTask' },
```

Comment:
`npx sb init` sets up `argTypesRegex` to apply actions to `on[A-Z]*`. Better to rely on that and keep the story definition simple?

Reply:
True, but that feels a bit too much like magic. I wanted to explain what's happening here.